File: software/examples/java/ExampleWriteReadType2.java
import com.tinkerforge.IPConnection;
import com.tinkerforge.BrickletNFCRFID;

public class ExampleWriteReadType2 {
	private static final String HOST = "localhost";
	private static final int PORT = 4223;

	// Change XYZ to the UID of your NFC/RFID Bricklet
	private static final String UID = "XYZ";

	// Note: To make the example code cleaner we do not handle exceptions. Exceptions
	//       you might normally want to catch are described in the documentation
	public static void main(String args[]) throws Exception {
		IPConnection ipcon = new IPConnection(); // Create IP connection
		// Note: Declare nr as final, so the listener can access it
		final BrickletNFCRFID nr = new BrickletNFCRFID(UID, ipcon); // Create device object

		ipcon.connect(HOST, PORT); // Connect to brickd
		// Don't use device before ipcon is connected

		// Add state changed listener
		nr.addStateChangedListener(new BrickletNFCRFID.StateChangedListener() {
			public void stateChanged(short state, boolean idle) {
				try {
					if(state == BrickletNFCRFID.STATE_REQUEST_TAG_ID_READY) {
						System.out.println("Tag found");

						// Write 16 byte to pages 5-8
						short[] dataWrite = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};

						nr.writePage(5, dataWrite);
						System.out.println("Writing data...");
					} else if(state == BrickletNFCRFID.STATE_WRITE_PAGE_READY) {
						// Request pages 5-8
						nr.requestPage(5);
						System.out.println("Requesting data...");
					} else if(state == BrickletNFCRFID.STATE_REQUEST_PAGE_READY) {
						// Get and print pages
						short[] data = nr.getPage();
						String s = "Read data: [" + data[0];

						for(int i = 1; i < data.length; i++) {
							s += " " + data[i];
						}

						s += "]";
						System.out.println(s);
					} else if((state & (1 << 6)) == (1 << 6)) {
						// All errors have bit 6 set
						System.out.println("Error: " + state);
					}

				} catch(Exception e) {
					System.out.println(e);
				}
			}
		});

		// Select NFC Forum Type 2 tag
		nr.requestTagID(BrickletNFCRFID.TAG_TYPE_TYPE2);

		System.out.println("Press key to exit"); System.in.read();
		ipcon.disconnect();
	}
}

File: software/examples/c/example_write_read_type2.c
#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_nfc_rfid.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change to your UID

// Callback function for state changed callback
void cb_state_changed(uint8_t state, bool idle, void *user_data) {
	NFCRFID *nfcrfid = (NFCRFID*)user_data;

	(void)idle; // avoid unused parameter warning

	if(state == NFC_RFID_STATE_REQUEST_TAG_ID_READY) {
		printf("Tag found\n");

		// Write 16 byte to pages 5-8
		uint8_t data_write[16] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};

		nfc_rfid_write_page(nfcrfid, 5, data_write);

		printf("Writing data...\n");
	} else if(state == NFC_RFID_STATE_WRITE_PAGE_READY) {
		// Request pages 5-8
		nfc_rfid_request_page(nfcrfid, 5);

		printf("Requesting data...\n");
	} else if(state == NFC_RFID_STATE_REQUEST_PAGE_READY) {
		uint8_t data_read[16];
		uint8_t i;

		// Get and print pages 5-8
		nfc_rfid_get_page(nfcrfid, data_read);
		printf("Read data: [%d", data_read[0]);

		for(i = 1; i < 16; i++) {
			printf(" %d", data_read[i]);
		}

		printf("]\n");
	} else if(state & (1 << 6)) {
		// All errors have bit 6 set
		printf("Error: %d\n", state);
	}
}

int main() {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	NFCRFID nfcrfid;
	nfc_rfid_create(&nfcrfid, UID, &ipcon);

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		exit(1);
	}
	// Don't use device before ipcon is connected

	// Register state changed callback to function cb_state_changed
	nfc_rfid_register_callback(&nfcrfid,
	                           NFC_RFID_CALLBACK_STATE_CHANGED,
	                           (void *)cb_state_changed,
	                           &nfcrfid);

	// Select NFC Forum Type 2 tag
	nfc_rfid_request_tag_id(&nfcrfid, NFC_RFID_TAG_TYPE_TYPE2);

	printf("Press key to exit\n");
	getchar();
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
}
